subsys/net/lib/nrf_cloud/src/nrf_cloud_logs_backend.c

```diff
@@ -0,0 +1,545 @@
+
+	log_output_func = log_format_func_t_get(log_format_current);
+	log_output_func(&log_nrf_cloud_output, &msg->log, log_output_flags);
+}
+
+static void logger_dropped(const struct log_backend *const backend, uint32_t cnt)
+{
+	if (backend == &log_nrf_cloud_backend) {
+		log_output_dropped_process(&log_nrf_cloud_output, cnt);
+		stats.lines_dropped += cnt;
+	}
+}
+
+static void logger_panic(const struct log_backend *const backend)
+{
+	if (backend == &log_nrf_cloud_backend) {
+		log_output_flush(&log_nrf_cloud_output);
+	}
+}
+
+static int logger_is_ready(const struct log_backend *const backend)
+{
+	static bool printed;
+
+	ARG_UNUSED(backend);
+	if (IS_ENABLED(CONFIG_NRF_CLOUD_MQTT)) {
+		if (nfsm_get_current_state() != STATE_DC_CONNECTED) {
+			return -EBUSY;
+		}
+	} else {
+		if (context.rest_ctx == NULL) {
+			if (rest_ctx) {
+				context.rest_ctx = rest_ctx;
+				LOG_WRN("Added rest_ctx");
+			}
+			if (!printed) {
+				printed = true;
+				LOG_WRN("rest_ctx is NULL");
+			}
+			return -EBUSY;
+		}
+	}
+
+	return 0; /* Logging is ready. */
+}
+
+static int logger_format_set(const struct log_backend *const backend, uint32_t log_type)
+{
+	if (backend != &log_nrf_cloud_backend) {
+		return 0;
+	}
+	if ((log_type == LOG_OUTPUT_TEXT) || (log_type == LOG_OUTPUT_DICT)) {
+		log_format_current = log_type;
+		return 0;
+	}
+	return -ENOTSUP;
+}
+
+static void logger_notify(const struct log_backend *const backend, enum log_backend_evt event,
+		       union log_backend_evt_arg *arg)
+{
+	if ((backend != &log_nrf_cloud_backend) ||
+	    (event != LOG_BACKEND_EVT_PROCESS_THREAD_DONE) ||
+	    (logger_is_ready(backend) != 0)) {
+		return;
+	}
+
+	/* Flush our transmission buffer */
+	send_ring_buffer();
+	LOG_INF("Logged lines:%u, bytes:%u; buf bytes:%u; "
+		"sent lines:%u, sent bytes:%u, dropped lines:%u",
+		stats.lines_rendered, stats.bytes_rendered,
+		ring_buf_size_get(&log_nrf_cloud_rb),
+		stats.lines_sent, stats.bytes_sent,
+		stats.lines_dropped);
+}
+
+static int send_ring_buffer(void)
+{
+	int err = 0;
+	int ret = 0;
+	struct nrf_cloud_tx_data output = {
+		.qos = MQTT_QOS_0_AT_MOST_ONCE,
+		.topic_type = (log_format_current == LOG_OUTPUT_TEXT) ?
+			       NRF_CLOUD_TOPIC_BULK : NRF_CLOUD_TOPIC_BIN
+	};
+	uint32_t stored;
+	uint8_t *base64_buf = NULL;
+
+	/* The bulk topic requires the multiple JSON messages to be placed in
+	 * a JSON array. Close the array then send it.
+	 */
+	if ((num_msgs != 0) && (log_format_current == LOG_OUTPUT_TEXT)) {
+		ring_buf_put(&log_nrf_cloud_rb, "]", 1);
+	}
+
+	stored = ring_buf_size_get(&log_nrf_cloud_rb);
+	output.data.len = ring_buf_get_claim(&log_nrf_cloud_rb,
+					     (uint8_t **)&output.data.ptr,
+					     stored);
+	if (output.data.len != stored) {
+		LOG_WRN("Capacity:%u, free:%u, stored:%u, claimed:%u",
+			ring_buf_capacity_get(&log_nrf_cloud_rb),
+			ring_buf_space_get(&log_nrf_cloud_rb),
+			stored, output.data.len);
+		stored = output.data.len;
+	}
```

Comment:
I've never seen this situation happen -- the values are always equal.  It could happen if we don't send the whole ring buffer each time, which could lead to valid data wrapping around.  But you're right, if we allow valid data to wrap, then the 2nd part of the valid data that starts at the beginning of the ring buffer memory would not get sent right away.  It would be sent on the next transfer, which would be OK for binary logs but not for JSON.